src/charm.py

```diff
@@ -13,92 +13,186 @@
+
+    def get_controller_ca(self) -> str:
+        """Get CA certificate used by targeted Juju controller.
+
+        CA certificate can be directly configured by `controller-ca` option, if it is, the value is
+        directly returned by this method. If it is not defined, a CA cert used by the controller
+        that deploys this unit will be returned.
         """
-        # Fetch the new config value
-        log_level = self.model.config["log-level"].lower()
-
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = BlockedStatus("invalid log level: '{log_level}'")
-
-    @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
-                }
-            },
-        }
+        explicit_cert = self.config.get("controller-ca", "")
+        if explicit_cert:
+            try:
+                return b64decode(explicit_cert, validate=True).decode(encoding="ascii")
+            except Base64Error as exc:
+                logger.error(
+                    "Config option 'controller-ca' does not contain valid base64-encoded data."
+                    " Bad data: %s",
+                    explicit_cert,
+                )
+                raise RuntimeError("Invalid base64 value in 'controller-ca' option.") from exc
+
+        agent_conf_path = pathlib.Path(hookenv.charm_dir()).joinpath("../agent.conf")
+        with open(agent_conf_path, "r", encoding="utf-8") as conf_file:
```

Comment:
What if open file fail? do we want to handle the exception?

Reply:
I don't think we can reasonably recover from such error. Maybe we can add an explicit log message but I'd let the the unit error-out anyway so that the hook/event execution is paused.